db/compaction/compaction_iterator.h

```diff
@@ -370,8 +368,36 @@ class CompactionIterator {
 
   // State
   //
+  enum ValidContext : uint8_t {
+    kMerge1 = 0,
+    kMerge2 = 1,
+    kParseKeyError = 2,
+    kCurrentKeyUncommitted = 3,
+    kKeepSDAndClearPut = 4,
+    kKeepHistory = 5,
+    kKeepSDForCC = 6,
+    kKeepSDForSnapshot = 7,
+    kKeepSD = 8,
+    kKeepDel = 9,
+    kNewUserKey = 10,
+  };
+  struct ValidityInfo {
+    inline bool IsValid() const { return rep & 1; }
+    ValidContext GetContext() const {
+      return static_cast<ValidContext>(rep >> 1);
+    }
+    inline void SetValid(uint8_t ctx) { rep = (ctx << 1) | 1; }
+    inline void Invalidate() { rep = 0; }
+    std::string ToString() const {
+      std::ostringstream oss;
+      oss << "valid=" << IsValid() << ", ctx=" << GetContext();
+      return oss.str();
+    }
+    uint8_t rep;
+  } validity_info_{.rep = 0};
```

Comment:
Designated initializers are a C++20 feature (we could use a default member initializer instead)